File: app/src/main/java/com/djacoronel/sleeklauncher/home/GridAdapter.java
package com.djacoronel.sleeklauncher.home;

import android.content.Context;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import com.djacoronel.sleeklauncher.R;

import com.djacoronel.sleeklauncher.data.model.AppDetail;
import java.util.List;

class GridAdapter extends RecyclerView.Adapter<GridAdapter.ViewHolder> {

  private List<AppDetail> apps;
  private Context mContext;

  GridAdapter(List<AppDetail> apps, Context context) {
    this.apps = apps;
    mContext = context;
  }

  class ViewHolder extends RecyclerView.ViewHolder {

    TextView label;
    ImageView icon;

    ViewHolder(View itemView) {
      super(itemView);
      label = itemView.findViewById(R.id.item_app_label);
      icon = itemView.findViewById(R.id.item_app_icon);
      itemView.setOnClickListener(
          new View.OnClickListener() {
            @Override
            public void onClick(View v) {
              ((MainActivity) mContext).launchApp(getAdapterPosition(), v);
            }
          }
      );
      itemView.setOnLongClickListener(
          new View.OnLongClickListener() {
            @Override
            public boolean onLongClick(View v) {
              // open launcher settings
              if (apps.get(getAdapterPosition()).getLabel().equals("Settings")) {
                ((MainActivity) mContext).openSettings();
              } else {
                ((MainActivity) mContext).iconLongClick(getAdapterPosition());
              }

              return false;
            }
          }
      );
    }
  }

  @Override
  public ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
    View v = LayoutInflater.from(parent.getContext())
        .inflate(R.layout.app_detail, parent, false);
    return new ViewHolder(v);
  }

  @Override
  public void onBindViewHolder(ViewHolder holder, int position) {
    holder.icon.setImageDrawable(apps.get(position).getIcon());
    holder.label.setText(apps.get(position).getLabel());
  }

  @Override
  public int getItemCount() {
    return apps.size();
  }
}

File: app/src/main/java/com/djacoronel/sleeklauncher/home/GridAdapter.kt
package com.djacoronel.sleeklauncher.home

import android.content.Context
import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import com.djacoronel.sleeklauncher.R
import com.djacoronel.sleeklauncher.data.model.AppDetail
import kotlinx.android.synthetic.main.app_detail.view.*

class GridAdapter(private val apps: List<AppDetail>, private val mContext: Context) : RecyclerView.Adapter<GridAdapter.ViewHolder>() {

    inner class ViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        fun bind(app: AppDetail) {
            itemView.item_app_icon.setImageDrawable(app.icon)
            itemView.item_app_label.text = app.label

            itemView.setOnClickListener { v ->
                if (app.label == "Sleek Launcher") (mContext as MainActivity).openSettings()
                else (mContext as MainActivity).launchApp(app, v)
            }

            itemView.setOnLongClickListener {
                (mContext as MainActivity).iconLongClick(app)
                false
            }
        }
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val v = LayoutInflater.from(parent.context)
                .inflate(R.layout.app_detail, parent, false)
        return ViewHolder(v)
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        holder.bind(apps[position])
    }

    override fun getItemCount(): Int = apps.size

    fun getApp(appActivityName: String): AppDetail {
        val app = apps.find { it.activity == appActivityName }
        return app!!
    }

    fun removeApp(appActivityName: String) {
        val app = apps.find { it.activity == appActivityName }
        val index = apps.indexOf(app)
        (apps as MutableList).removeAt(index)
        notifyItemRemoved(index)
    }
}
